File: crunchy-core/src/main/java/io/crunchy/core/stage/DuplicateDetectionStage.java
package io.crunchy.core.stage;

import io.crunchy.core.pipeline.Block;
import io.crunchy.core.pipeline.PipelineStage;
import io.crunchy.core.pipeline.ProcessingContext;
import io.crunchy.core.text.TextSimilarity;

import java.util.ArrayList;
import java.util.List;

/**
 * Stage 3: exact and near-duplicate blocks collapse into the first occurrence,
 * which records how often the point was repeated (a useful emphasis signal for
 * ranking).
 */
public final class DuplicateDetectionStage implements PipelineStage {

    @Override
    public String name() {
        return "duplicate-detection";
    }

    @Override
    public void process(ProcessingContext context) {
        List<Block> kept = new ArrayList<>();
        int duplicates = 0;
        for (Block block : context.getBlocks()) {
            if (block.isDropped()) {
                continue;
            }
            Block original = findDuplicate(kept, block);
            if (original != null) {
                original.incrementDuplicates();
                block.drop("duplicate of " + original.getOrigin());
                duplicates++;
            } else {
                kept.add(block);
            }
        }
        if (duplicates > 0) {
            context.getIgnoredContent().add(duplicates + " duplicate/near-duplicate paragraph(s) collapsed");
        }
    }

    private Block findDuplicate(List<Block> kept, Block candidate) {
        for (Block existing : kept) {
            if (TextSimilarity.nearDuplicate(existing.getText(), candidate.getText())) {
                return existing;
            }
        }
        return null;
    }
}

File: web/src/core/stage/DuplicateDetectionStage.ts
import { Block } from '../pipeline/Block.js';
import { PipelineStage } from '../pipeline/PipelineStage.js';
import { ProcessingContext } from '../pipeline/ProcessingContext.js';
import { nearDuplicate } from '../text/TextSimilarity.js';

/**
 * Stage 3: exact and near-duplicate blocks collapse into the first occurrence,
 * which records how often the point was repeated (a useful emphasis signal for
 * ranking).
 */
export class DuplicateDetectionStage implements PipelineStage {
  readonly name = 'duplicate-detection';

  process(context: ProcessingContext): void {
    const kept: Block[] = [];
    let duplicates = 0;
    for (const block of context.blocks) {
      if (block.dropped) {
        continue;
      }
      const original = this.findDuplicate(kept, block);
      if (original) {
        original.incrementDuplicates();
        block.drop(`duplicate of ${original.origin}`);
        duplicates++;
      } else {
        kept.push(block);
      }
    }
    if (duplicates > 0) {
      context.ignoredContent.push(`${duplicates} duplicate/near-duplicate paragraph(s) collapsed`);
    }
  }

  private findDuplicate(kept: Block[], candidate: Block): Block | undefined {
    return kept.find((existing) => nearDuplicate(existing.text, candidate.text));
  }
}
